### PR TITLE
refactor(ingest): Call source_helpers via new WorkUnitProcessors in base Source

### PR DESCRIPTION
**Pull request by asikowitz**
Somewhat serious refactor that provides a default implementation of `get_workunits` and changes `get_workunits_internal` to be the main overriden method. Adds `get_workunit_processors()` which returns a list of callables that take in a `WorkUnit` stream and alter that stream (i.e. source helpers). Provides a default list of workunit processors (source helpers) that can be changed by overriding the method.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/api/source_helpers.py [hunk 1]
```diff
@@ -119,14 +118,9 @@ def auto_workunit_reporter(report: SourceReport, stream: Iterable[T]) -> Iterabl
 
 def auto_materialize_referenced_tags(
     stream: Iterable[MetadataWorkUnit],
-    active: bool = True,
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/api/source_helpers.py:122]: Now handled by passing in `None` for the workunit processor, although I don't think this was ever passed as `False`

File: metadata-ingestion/src/datahub/ingestion/source/dbt/dbt_common.py [hunk 2]
```diff
@@ -688,12 +693,8 @@ def __init__(self, config: DBTCommonConfig, ctx: PipelineContext, platform: str)
                 self.config.owner_extraction_pattern
             )
         # Create and register the stateful ingestion use-case handler.
-        self.stale_entity_removal_handler = StaleEntityRemovalHandler(
-            source=self,
-            config=self.config,
-            state_type_class=GenericCheckpointState,
-            pipeline_name=self.ctx.pipeline_name,
-            run_id=self.ctx.run_id,
+        self.stale_entity_removal_handler = StaleEntityRemovalHandler.create(
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/dbt/dbt_common.py:696]: This gets used to call `self.stale_entity_removal_handler.add_urn_to_skip(node_datahub_urn)` at some point

File: metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py [hunk 3]
```diff
@@ -1197,6 +1185,17 @@ def get_workspace_workunit(
                 for workunit in dataset_workunits:
                     yield workunit
 
+    def get_workunit_processors(self) -> Sequence[Optional[MetadataWorkUnitProcessor]]:
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/powerbi/powerbi.py:1188]: Pay attention to this file, some pretty non-standard behavior. Not sure if there's a cleaner way to do this... could directly overwrite `get_workunits` as one option

File: metadata-ingestion/src/datahub/ingestion/source/file.py [hunk 4]
```diff
@@ -205,8 +207,8 @@ def get_filenames(self) -> Iterable[str]:
             self.report.total_num_files = 1
             return [str(self.config.path)]
 
-    def get_workunits(self) -> Iterable[MetadataWorkUnit]:
-        return auto_workunit_reporter(self.report, self.get_workunits_internal())
+    def get_workunit_processors(self) -> Sequence[Optional[MetadataWorkUnitProcessor]]:
+        return [partial(auto_workunit_reporter, self.report)]
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/file.py:211]: add a comment that not calling super() is intentional here

File: metadata-ingestion/src/datahub/ingestion/api/source.py [hunk 5]
```diff
@@ -155,9 +176,35 @@ def create(cls, config_dict: dict, ctx: PipelineContext) -> "Source":
         # can't make this method abstract.
         raise NotImplementedError('sources must implement "create"')
 
-    @abstractmethod
+    def get_workunit_processors(self) -> Sequence[Optional[MetadataWorkUnitProcessor]]:
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/api/source.py:179]: I think we should change this return type to `List[Optional[MetadataWorkUnitProcessor]]`

File: metadata-ingestion/tests/integration/integration_helpers.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+from datahub.ingestion.source.state.checkpoint import Checkpoint
+from datahub.ingestion.source.state.stale_entity_removal_handler import (
+    StaleEntityRemovalHandler,
+)
+from datahub.ingestion.source.state.stateful_ingestion_base import (
+    StatefulIngestionSourceBase,
+)
+
+
+def get_current_checkpoint_from_pipeline(
```
Comment (hsheth2) [inline metadata-ingestion/tests/integration/integration_helpers.py:13]: we're gonna get merge conflicts on this with my PR

File: metadata-ingestion/src/datahub/ingestion/api/source.py (hunk 5 above)
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/api/source.py:179]: So we can `append` to it? I guess it's more specific.

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/api/source.py:179]: not for append - we shouldn't be using append here

Sequence is a bit weird to work with in some places with mypy. We can always make it more general in the future, but let's not restrain ourselves unnecessarily

Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/api/source.py:179]: Oh really, what issues have you seen with Sequence? In general, I think we want to use `Sequence` (if possible) for parameters but `List` / other specific types for return types

Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/api/source.py:179]: Actually i think sequences are fine. iirc the issue i've seen is that Iterable is not covariant (so passing Iterable[Square] to a function that takes Iterable[Shape] would throw an error), but sequence works fine there

> In general, I think we want to use Sequence (if possible) for parameters but List / other specific types for return types

yup this seems like a good rule of thumb - might be worth adding to the list here https://datahubproject.io/docs/metadata-ingestion/developing/#code-style

File: metadata-ingestion/tests/integration/integration_helpers.py (hunk 6 above)
Comment (hsheth2) [reply metadata-ingestion/tests/integration/integration_helpers.py:13]: ref https://github.com/datahub-project/datahub/pull/8104

**Review by hsheth2 — APPROVED** [review]
LGTM

**Review by mayurinehate — APPROVED** [review]
LGTM

File: metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py [hunk 7]
```diff
@@ -491,6 +474,14 @@ def gen_dataset_containers(
             tags=tags_joined,
         )
 
+    def get_workunit_processors(self) -> List[Optional[MetadataWorkUnitProcessor]]:
+        return [
+            *super().get_workunit_processors(),
+            StaleEntityRemovalHandler.create(
+                self, self.config, self.ctx
+            ).workunit_processor,
+        ]
```
[the remark this answers is not in the document]
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/bigquery.py:483]: Yeah, was trying to match existing behavior. If we want it the other way we can always make the change

File: metadata-ingestion/src/datahub/ingestion/api/source.py [hunk 8]
```diff
@@ -155,9 +177,35 @@ def create(cls, config_dict: dict, ctx: PipelineContext) -> "Source":
         # can't make this method abstract.
         raise NotImplementedError('sources must implement "create"')
 
-    @abstractmethod
+    def get_workunit_processors(self) -> List[Optional[MetadataWorkUnitProcessor]]:
+        """A list of functions that transforms the workunits produced by this source.
```
[the remark this answers is not in the document]
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/api/source.py:181]: Yeah, the idea is we can do something like `return [*super().get_workunit_processors(), other_workunit_processor if self.config.flag else None]`

**Pull request merged**